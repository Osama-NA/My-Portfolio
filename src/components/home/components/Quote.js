import React, { useEffect, useState } from 'react';
import {QuoteContainer} from '../styles/Quote.styled';

const LEARN = "Learn";
const DEVELOP = "Develop";
const MAC = "Make a change";

const size = (str) => str.length;

export const Quote = () => {

    const [learn, setLearn] = useState("");
    const [develop, setDevelop] = useState("");
    const [mac, setMac] = useState("");
    const [learnIndex, setLearnIndex] = useState(0);
    const [developIndex, setDevelopIndex] = useState(0);
    const [macIndex, setMacIndex] = useState(0);
    const [index, setIndex] = useState(0);

    // on render type text at speed 100ms/s
    useEffect(() => setTimeout(typeText, 100));

    // Types text based on index state conditions 
    const typeText = () => {
        if (index < size(LEARN)){
            typeLearn();
        } else if (index < size(LEARN) + size(DEVELOP) && index >= size(LEARN)){
            typeDevelop();
        } else if (index < size(LEARN) + size(DEVELOP) + size(MAC) && index >= size(LEARN) + size(DEVELOP) ){
            typeMac();
        } 
        setIndex(index + 1);
    }

    // These functions type the text letter by letter and increase the index of each text respectively
    const typeLearn = () => {
        setLearn(LEARN.substring(0, learnIndex));
        setLearnIndex(learnIndex + 1);
    }

    const typeDevelop = () => {
        setDevelop(DEVELOP.substring(0, developIndex));
        setDevelopIndex(developIndex + 1);
    }

    const typeMac = () => {
        setMac(MAC.substring(0, macIndex));
        setMacIndex(macIndex + 1);
    }

    //These functions return display style of span elements (".", "_")
    const getLearnDisplay = () => {
        return size(learn) === size(LEARN) ? {display:"inline"} : {display: "none"} ;
    }

    const getDevelopDisplay = () => {
        return size(develop) === size(DEVELOP) ? {display:"inline"} : {display: "none"} ;
    }

    const getMacDisplay = () => {
        return size(mac) === size(MAC) ? {display:"inline"} : {display: "none"} ;
    }

    // returns whether to show or not show typing cursor for the given text
    const getCursorCondition = (text) => {
        switch(text){
            case LEARN:
                return index < size(LEARN) ? "_" : "";
            case DEVELOP:
                return index < size(LEARN) + size(DEVELOP) && index >= size(LEARN) ? "_" : "";
            case MAC:
                return index < size(LEARN) + size(DEVELOP) + size(MAC) && index >= size(LEARN) + size(DEVELOP) ? "_" : "";
            default:
                return;
        }
    }

    return (
        <QuoteContainer>
            <p className="learn">
                {learn}{getCursorCondition(LEARN)}
                <span style={getLearnDisplay()}>.</span>
            </p>
            <p className="develop">
                {develop}{getCursorCondition(DEVELOP)}
                <span style={getDevelopDisplay()}>.</span>
            </p>
            <p className="mac">
                {mac}{getCursorCondition(MAC)}
                <span style={getMacDisplay()}>_</span>
            </p>
        </QuoteContainer>
    )
}
