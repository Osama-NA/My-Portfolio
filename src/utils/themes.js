// GREEN / GRAY
export const greenGrayTheme = {
    themeColors: {
        mainColor: '#54EBA0'
    },
    fontColors: {
        secondary: '#e0e0e0',
        primary: '#fff',
        gray: '#5e5e5e'
    },
    backgroundColors: {
        lightSide: '#313138',
        darkSide: '#242429',
        lightBackground: '#3F3F47',
        navBackground: '#383840'
    },
    cubesColor: '#03AD78',
    circleBackground: '#313138'
}

// BLUEISH PURPLE / GRAY 
export const blueishPurpleGrayTheme = {
    themeColors: {
        mainColor: '#674DFF'
    },
    fontColors: {
        secondary: '#e0e0e0',
        primary: '#fff',
        gray: '#5e5e5e'
    },
    backgroundColors: {
        lightSide: '#313138',
        darkSide: '#242429',
        lightBackground: '#3F3F47',
        navBackground: '#383840'
    },
    cubesColor: '#544DFF',
    circleBackground: '#313138'
}

// PURPLE / GRAY
export const purpleGrayTheme = {
    themeColors: {
        mainColor: '#B15EFF'
    },
    fontColors: {
        secondary: '#e0e0e0',
        primary: '#fff',
        gray: '#5e5e5e'
    },
    backgroundColors: {
        lightSide: '#313138',
        darkSide: '#242429',
        lightBackground: '#3F3F47',
        navBackground: '#383840'
    },
    cubesColor: '#9A52DE',
    circleBackground: '#313138'
}

// LIGHT BLUE / GRAY
export const lightBlueGrayTheme = {
    themeColors: {
        mainColor: '#00E6E2'
    },
    fontColors: {
        secondary: '#e0e0e0',
        primary: '#fff',
        gray: '#5e5e5e'
    },
    backgroundColors: {
        lightSide: '#313138',
        darkSide: '#242429',
        lightBackground: '#3F3F47',
        navBackground: '#383840'
    },
    cubesColor: '#00E6E2',
    circleBackground: '#313138'
}

// YELLOW / GRAY
export const yellowGrayTheme = {
    themeColors: {
        mainColor: '#F2F270'
    },
    fontColors: {
        secondary: '#e0e0e0',
        primary: '#fff',
        gray: '#5e5e5e'
    },
    backgroundColors: {
        lightSide: '#313138',
        darkSide: '#242429',
        lightBackground: '#3F3F47',
        navBackground: '#383840'
    },
    cubesColor: '#F2F208',
    circleBackground: '#313138'
}

// PINK / GRAY
export const pinkGrayTheme = {
    themeColors: {
        mainColor: '#FF558E'
    },
    fontColors: {
        secondary: '#e0e0e0',
        primary: '#fff',
        gray: '#5e5e5e'
    },
    backgroundColors: {
        lightSide: '#313138',
        darkSide: '#242429',
        lightBackground: '#3F3F47',
        navBackground: '#383840'
    },
    cubesColor: '#FF558E',
    circleBackground: '#313138'
}

// RED / GRAY
export const redGrayTheme = {
    themeColors: {
        mainColor: '#FF3142'
    },
    fontColors: {
        secondary: '#e0e0e0',
        primary: '#fff',
        gray: '#5e5e5e'
    },
    backgroundColors: {
        lightSide: '#313138',
        darkSide: '#242429',
        lightBackground: '#3F3F47',
        navBackground: '#383840'
    },
    cubesColor: '#FF3142',
    circleBackground: '#313138'
}