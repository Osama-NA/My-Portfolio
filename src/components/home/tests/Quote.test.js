import React from 'react';
import ReactDOM from 'react-dom';
import renderer from 'react-test-renderer';
import { ThemeProvider } from 'styled-components';
import { BrowserRouter } from 'react-router-dom';
import { Quote } from '../components/Quote.js';
import { cleanup, render } from '@testing-library/react';
import { greenGrayTheme } from '../../../utils/themes.js';

afterEach(cleanup);

describe("Quote component tests", () => {

    const QuoteComponent = () => {
        return (
            <BrowserRouter>
                <ThemeProvider theme={greenGrayTheme}>
                    <Quote />
                </ThemeProvider>
            </BrowserRouter>
        )
    }

    it("Quote component renders without crashing", () => {
        let div = document.createElement('div');
        ReactDOM.render(<QuoteComponent />, div);
    })

    it("Quote component is in the document", () => {
        const { getByTestId } = render(<QuoteComponent />);
        expect(getByTestId('quote')).toBeInTheDocument();
    })

    it("Quote component matches snapshot", () => {
        const tree = renderer.create(<QuoteComponent />).toJSON();
        expect(tree).toMatchSnapshot();
    })
})