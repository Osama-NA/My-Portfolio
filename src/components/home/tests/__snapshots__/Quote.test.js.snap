// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Quote component tests Quote component matches snapshot 1`] = `
<div
  className="sc-bdfBQB duMjGC"
  data-testid="quote"
>
  <i
    className="fas fa-quote-left quote"
  />
  <p
    className="learn"
  >
    
    _
    <span
      style={
        Object {
          "display": "none",
        }
      }
    >
      .
    </span>
  </p>
  <p
    className="develop"
  >
    
    
    <span
      style={
        Object {
          "display": "none",
        }
      }
    >
      .
    </span>
  </p>
  <p
    className="mac"
  >
    
    
    <span
      style={
        Object {
          "display": "none",
        }
      }
    >
      .
    </span>
  </p>
</div>
`;
